mongo/data_store.go

```diff
@@ -166,6 +166,66 @@ func (m *JobStore) GetTask(ctx context.Context, jobID string, taskName string) (
 	return task, err
 }
 
+// GetTasks retrieves all the tasks, from the collection, and lists them in order of last_updated
+func (m *JobStore) GetTasks(ctx context.Context, offsetParam string, limitParam string, jobID string) (models.Tasks, error) {
+	s := m.Session.Copy()
+	defer s.Close()
+	log.Info(ctx, "getting list of tasks")
+	results := models.Tasks{}
+
+	// If an empty jobID or taskName was passed in, return an error with a message.
+	if jobID == "" {
+		return results, ErrEmptyIDProvided
+	}
+
+	_, err := m.findJob(s, jobID, models.Job{})
+	if err != nil {
+		if err == mgo.ErrNotFound {
+			return results, ErrJobNotFound
+		}
+		return results, err
+	}
+
+	numTasks, _ := s.DB(m.Database).C(m.TasksCollection).Find(bson.M{"job_id": jobID}).Count()
+	log.Info(ctx, "number of tasks found in tasks collection", log.Data{"numTasks": numTasks})
+
+	if numTasks == 0 {
+		log.Info(ctx, "there are no tasks in the data store - so the list is empty")
+		results.TaskList = make([]models.Task, 0)
+		return results, nil
+	}
+
+	// Get all the tasks from the tasks collection and order them by lastupdated
+	iter := s.DB(m.Database).C(m.TasksCollection).Find(bson.M{"job_id": jobID}).Sort("last_updated").Iter()
+	defer func() {
+		err := iter.Close()
+		if err != nil {
+			log.Error(ctx, "error closing iterator", err)
+		}
+	}()
+
+	tasks := make([]models.Task, numTasks)
+	if err := iter.All(&tasks); err != nil {
+		return results, err
+	}
+
+	paginator := pagination.NewPaginator(m.cfg.DefaultLimit, m.cfg.DefaultOffset, m.cfg.DefaultMaxLimit)
```

Comment:
I would have moved the pagination validation before requesting data from mongo, if validation fails because of bad offset, limit values then return before making requests to the database. This will be a performance gain (more efficient).

Once validation occurs then call into the database using this method, you will need to pass in the offset and limit values into this method.

Again use the mongo db querying language to retrieve data instead of iterating all documents and working it out below as this again increases memory usage and will be slower to respond.
The above can be done using [Skip](https://docs.mongodb.com/compass/current/query/skip/) (represents offset); [Limit](https://docs.mongodb.com/compass/current/query/limit/) (represents offset) and [OrderBy](https://docs.mongodb.com/manual/reference/operator/meta/orderby/) (represent sort).

Reply:
Thanks for this suggestion - it's good to know about those mongo db querying language methods. I've made those changes now.